tools/start-local-vm

```diff
@@ -0,0 +1,242 @@
+        case ${image} in
+            *data*) readonly data_image=${image} ;;
+            *)      readonly boot_image=${image} ;;
+        esac
+    done
+}
+
+create_extra_files() {
+    # Explicitly instruct the kernel to send its output to the serial port on
+    # x86 via a bootconfig initrd. Passing in settings via user-data would be
+    # too late to get console output of the first boot.
+    if [[ ${arch} = x86_64 ]]; then
+        extra_files["${repo_root}/tools/bootconfig/qemu-x86-console-bootconfig.data"]=bootconfig.data
+    fi
+}
+
+inject_files() {
+    if [[ ${#extra_files[@]} -eq 0 ]]; then
+        return 0
+    fi
+
+    local loopback_device
+    if ! loopback_device=$(sudo losetup --find --show --partscan "${boot_image}"); then
+        bail 'Failed to create the loopback device needed to splice in user data.'
+    fi
+
+    # Finding the partition may need a few attempts while the partition scan completes
+    local private_partition
+    for (( i = 0; i < 10; i++ )); do
+        private_partition=$(lsblk --list --output NAME,PARTLABEL "${loopback_device}" \
+            | awk '$2 == "BOTTLEROCKET-PRIVATE" { print $1 }')
+        [[ -n "${private_partition}" ]] && break
+        sleep 0.1
+    done
+    private_partition=/dev/${private_partition}
+
+    if [[ ! -b ${private_partition} ]]; then
+        sudo losetup --detach "${loopback_device}"
+        bail 'Failed to find the private partition.'
+    fi
+
+    local private_mount
+    private_mount=$(mktemp -d)
+    sudo mount "${private_partition}" "${private_mount}"
+    sudo chown "${USER}" "${private_mount}"
+
+    for local_file in "${!extra_files[@]}"; do
+        local image_file=${extra_files[${local_file}]}
+        sudo cp "${local_file}" "${private_mount}/${image_file}"
+    done
+
+    sudo umount "${private_partition}"
+    rm -r "${private_mount}"
+
+    sudo losetup --detach "${loopback_device}"
+}
+
+launch_vm() {
+    local -a qemu_args=(
+        -nographic
+        -enable-kvm
+        -cpu host
+        -smp "${vm_cpus}"
+        -m "${vm_mem}"
+        -drive index=0,if=virtio,format=raw,file="${boot_image}"
+    )
+
+    # Plug the virtual primary NIC in as BDF 00:10.0 so udev will give it a
+    # consistent name we can know ahead of time--enp0s16 or ens16.
+    qemu_args+=( -netdev user,id=net0,hostfwd=tcp::2222-:22 -device virtio-net-pci,netdev=net0,addr=10.0 )
+
+    # Resolve the last bit of uncertainty by disabling ACPI-based PCI hot plug,
+    # causing udev to use the bus location when naming the NIC (enp0s16). Since
+    # QEMU does not support PCI hot plug via ACPI on Arm, turn it off for the
+    # emulated x86_64 chipset only to achieve parity.
+    if [[ ${arch} = x86_64 ]]; then
+        qemu_args+=( -global PIIX4_PM.acpi-root-pci-hotplug=off )
+    fi
+
+    if [[ ${arch} = aarch64 ]]; then
+        qemu_args+=( -machine virt )
+        qemu_args+=( -bios /usr/share/edk2/aarch64/QEMU_EFI.silent.fd )
```

Comment:
Is this provided by all distros? Or is distro specific, and the file is installed in different locations depending on the distro?

Reply:
Totally decided by a distro's packagers. On current Fedora, this is where the `edk2-arm` package (pulled in by `qemu`) puts it. On an Ubuntu machine it might be provided by `ovmf` and only pulled in as a recommended package. For a start, I wanted to go with the fixed location and see whether there's actually any breakage.